File: server/src/core/wsHub.ts
import type WebSocket from "ws";

/**
 * WS 连接中心：
 * - 维护 userId -> WebSocket 的全局映射
 * - 统一封装推送与连接回收逻辑，避免各模块重复实现
 */

export const onlineWsByUserId = new Map<number, WebSocket>();

/** 绑定（或替换）某个用户的连接 */
export function bindUserSocket(userId: number, ws: WebSocket) {
  const old = onlineWsByUserId.get(userId);
  if (old && old !== ws) {
    try {
      old.close(1000, "Replaced by new connection");
    } catch {}
  }
  onlineWsByUserId.set(userId, ws);
}

/** 解绑连接（仅当当前连接仍是该用户的活跃连接时才删除） */
export function unbindUserSocket(userId: number, ws: WebSocket) {
  const current = onlineWsByUserId.get(userId);
  if (current === ws) {
    onlineWsByUserId.delete(userId);
  }
}

/** 判断用户是否在线 */
export function isUserOnline(userId: number): boolean {
  return onlineWsByUserId.has(userId);
}

/** 向指定用户推送 JSON（用户离线或推送失败时返回 false） */
export function pushJsonToUser(userId: number, payload: unknown): boolean {
  const ws = onlineWsByUserId.get(userId);
  if (!ws) return false;
  try {
    ws.send(JSON.stringify(payload));
    return true;
  } catch {
    return false;
  }
}

File: java-server/src/main/java/com/todothink/core/ws/WsHub.java
package com.todothink.core.ws;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;
import org.springframework.web.socket.CloseStatus;
import org.springframework.web.socket.TextMessage;
import org.springframework.web.socket.WebSocketSession;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

/**
 * WebSocket 连接中心：维护 userId -> WebSocketSession 映射，统一推送与连接回收。
 */
@Component
public class WsHub {

    private static final Logger log = LoggerFactory.getLogger(WsHub.class);

    private final ConcurrentHashMap<Integer, WebSocketSession> onlineWsByUserId = new ConcurrentHashMap<Integer, WebSocketSession>();
    private final ObjectMapper objectMapper;

    public WsHub(ObjectMapper objectMapper) {
        this.objectMapper = objectMapper;
    }

    public void bindUserSocket(int userId, WebSocketSession session) {
        WebSocketSession old = onlineWsByUserId.get(userId);
        if (old != null && old != session && old.isOpen()) {
            try {
                old.close(CloseStatus.NORMAL.withReason("Replaced by new connection"));
            } catch (Exception e) {
                log.debug("close old ws failed userId={}", userId, e);
            }
        }
        onlineWsByUserId.put(userId, session);
    }

    public void unbindUserSocket(int userId, WebSocketSession session) {
        WebSocketSession current = onlineWsByUserId.get(userId);
        if (current == session) {
            onlineWsByUserId.remove(userId);
        }
    }

    public boolean isUserOnline(int userId) {
        return onlineWsByUserId.containsKey(userId);
    }

    public boolean pushJsonToUser(int userId, Map<String, Object> payload) {
        WebSocketSession session = onlineWsByUserId.get(userId);
        if (session == null || !session.isOpen()) {
            return false;
        }
        try {
            String json = objectMapper.writeValueAsString(payload);
            synchronized (session) {
                session.sendMessage(new TextMessage(json));
            }
            return true;
        } catch (Exception e) {
            log.debug("push ws failed userId={}", userId, e);
            return false;
        }
    }
}
